client/concordclient/include/client/concordclient/concord_client.hpp

```diff
@@ -130,14 +132,59 @@ class ConcordClient {
 
   // Register a callback that gets invoked for every validated event received.
   // void callback(EventGroup);
+  // Return subscriber ID used to unsubscribe.
   template <class CallbackT>
   void subscribe(const SubscribeRequest& request,
```

Comment:
The callback seems to have a very specific signature, it accepts EventGroups. I do not see the point of making the method templated. I suggest passing callback as `std::function` with a clear list of arguments.

Additionally, once we get rid of templates, we may move the implementation to a CPP file to decrease the compilation time and improve incapslation.

Reply:
The OP at SO is concerned with performance, which is not the problem in our case: subscription does not happen too often. Regarding the compile-time guarantees, `std::function` provides them: if you try to pass a function with the wrong signature, the compiler will fail.

Reply:
Performance is probably no issue for `subscribe` but might be for `send` and the APIs should look the same IMHO. Is your only objection that the callback parameter are defined in the documentation and not in the code or is there more?

Reply:
We talked offline. TLDR: We believe that the potential performance boost of templates will be negligible at runtime and hence we choose `std::function` for practicality. 

Our conclusion is that we don't see the callback invocation itself as a performance bottleneck for now. For both APIs ,`send` and `subscribe`, concord client is going to wait on multiple replicas to respond and compare the results. Invoking the callback via templates or std::function afterwards seems negligible in comparison. However, we also have no data (we should do this once concord client is fully implemented). We agreed that `std::function` makes the interface straightforward and easier to maintain. Templates require all functions to be defined in header files and callback types to be "defined" in comments. If performance is not an issue then we prefer a clean interface via `std::function`.